### DISCUSSION TITLE
"Fatal: Could not detach HEAD" Error During Update MOOSE Repository

### DISCUSSION THREAD

**Discussion by styyokuda** [idaholab/moose#27797] [Q&A Getting Started] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Issue or question about MOOSE

All,
I got the error message during updating MOOSE repository as:
~/projects/moose$ git rebase origin/master
First, rewinding head to replay your work on top of it...
error: Updating the following directories would lose untracked files in them:
	modules/tensor_mechanics/src
Aborting
fatal: Could not detach HEAD

It seems this kind of error was discussed in "Fatal" warning during MOOSE update #26465.
I have the "tensor_mechanics" in the module directory and the "update_and_rebuild_wasp.sh" the scrips directory.  What am I supposed to do then? Please help. S. Thomas

### Diagnostics for MOOSE installation issues

[Optional] Output of the diagnostics scripts. To run the script: cd ~/projects/moose/scripts; ./diagnostics.sh

**Comment by GiudGiud** [comment]
Hello

We deleted that module.
Move any work you have in that module into the solid mechanics module. Either manually, or with git mv in the first commit adding the new files
Guillaume

**Reply by styyokuda** [reply]
Thank you, Guillaume.
Now, with "conda create -n moose moose-dev=2024.05.13"
The action shown in the attached has been going on overnight.
I don't know what to do. Please help.

![Screenshot 2024-06-06 at 8 36 10 AM](https://github.com/idaholab/moose/assets/77596591/a4866b5d-9096-4f0b-ad5f-c0d5891ff0db)

**Reply by milljm** [reply]
This usually happens when your instance of Conda is using the deprecated solver. Newer versions are using libmamba. Can you cancel that (`ctrl-c`), and run the following:

```bash
conda list | grep conda
```
Once we see whats there, we can continue appropriately.

**Reply by styyokuda** [reply]
Thank you. Please see below.
![Screenshot 2024-06-06 at 11 04 30 AM](https://github.com/idaholab/moose/assets/77596591/cb843493-74ac-4499-969a-d06b88576d37)

**Reply by milljm** [reply]
Sorry for the delay... Yeah it looks like you are running an old version of Conda. If Conda will let you, try to upgrade it:
```bash
conda update conda
```

If you do not see conda being updated to 23.11.x, we will need to be a bit more aggressive with it. Which is usually the case. If so, you can try to force a newer version with:
```bash
conda install conda=25
```

Once you have a version >=23.11, then try to create the moose environment:
```bash
conda create -n moose moose-dev=2024.05.11
```

**Reply by styyokuda** [reply]
So, I did "conda update conda" (see below)
![Screenshot 2024-06-06 at 2 14 41 PM](https://github.com/idaholab/moose/assets/77596591/befcd776-a951-4c9c-8b3c-1a417720e9be)
![Screenshot 2024-06-06 at 2 15 51 PM](https://github.com/idaholab/moose/assets/77596591/255ddc25-39b0-429b-aa20-e04ca315cd74)

But, "grep list | grep conda" shows still 23.1.0 (see below)
![Screenshot 2024-06-06 at 2 16 26 PM](https://github.com/idaholab/moose/assets/77596591/14054573-da2d-421b-91ca-8f13e22cb141)
So, I did "conda install conda=25" and got the following

![Screenshot 2024-06-06 at 2 17 16 PM](https://github.com/idaholab/moose/assets/77596591/49747697-09eb-4235-ba2c-fab8bebc571b)

Also, I tried "conda install conda=24.5.0" as suggested in the first printout above and it did something (see the last part of printout)
![Screenshot 2024-06-06 at 2 29 04 PM](https://github.com/idaholab/moose/assets/77596591/97bdbc63-2998-4933-b88c-9abad0394086)

BNut  "grep list | grep conda" show the same 23.1.0 (see below).
![Screenshot 2024-06-06 at 2 30 56 PM](https://github.com/idaholab/moose/assets/77596591/ab2eab67-b712-4adf-94ac-899a326788b1)

**Reply by GiudGiud** [reply]
I would advise to just delete the conda directory and download it again
then `conda update conda` should work

**Reply by milljm** [reply]
Also, if you can, please do not post images of text. WSL text windows are capable of being copy pasted like any other terminal window; mouse select, right-click usually does the trick.

**Reply by styyokuda** [reply]
Do you mean do "rm -r conda" in the /projects/moose directory 
and follow the instructions of Conda MOOSE Environment of installing MOOSE from Getting Started
(install to my home directory - not in the /projects/moose directory)?
Please let me know.

**Reply by GiudGiud** [reply]
If you put the (mini)conda directory there then yes

We expect that (mini)conda directory to be under the home directory iirc

**Reply by milljm** [reply]
With the base environment loaded:
```bash
conda deactivate
conda activate base
```
You can get the installation directory by then running:
```bash
echo $CONDA_PREFIX
```
This will be the directory you want to remove.

We may run into an initialization issue after installing a newer version of Conda, but we will deal with that when/if it happens. Regardless, once you remove the directory above, please follow our [Conda Installation Instructions](https://mooseframework.inl.gov/getting_started/installation/conda.html) to obtain the latest version available.

And if installation proceeds well, next follow the section of Installing MOOSE (`conda create -n moose moose-dev=2024.05.13`).

**Reply by styyokuda** [reply]
Thank  you all for your help. So I did the following and got stuck at "conda init --all."  Pease help.

(base) d3k768@WE47574 ~ % conda deactivate
d3k768@WE47574 ~ % conda activate base
(base) d3k768@WE47574 ~ % echo $CONDA_PREFIX
/Users/d3k768/mambaforge3
(base) d3k768@WE47574 ~ % rm -r  /Users/d3k768/mambaforge3
override r--r--r-- d3k768/staff for /Users/d3k768/mambaforge3/pkgs/perl-5.32.1-4_hf2054a2_perl5/info/licenses/Copying? y
override r--r--r-- d3k768/staff for /Users/d3k768/mambaforge3/pkgs/perl-5.32.1-4_hf2054a2_perl5/info/licenses/Artistic? y
override r--r--r-- d3k768/staff for /Users/d3k768/mambaforge3/pkgs/perl-5.32.1-2_h9b22ae9_perl5/info/licenses/Copying? y
override r--r--r-- d3k768/staff for /Users/d3k768/mambaforge3/pkgs/perl-5.32.1-2_h9b22ae9_perl5/info/licenses/Artistic? y
.
.
(base) d3k768@WE47574 ~ % export PATH=$HOME/miniforge/bin:$PATH
(base) d3k768@WE47574 ~ % conda init --all
exit
__conda_exe:1: no such file or directory: /Users/d3k768/mambaforge3/bin/conda

Saving session...
...copying shared history...cp: /Users/d3k768/.zsh_history: Permission denied

...saving history...shell_session_save_history:14: permission denied: /Users/d3k768/.zsh_history
truncating history files...shell_session_save_history:21: can't read history file /Users/d3k768/.zsh_history

[Process completed]

**Reply by GiudGiud** [reply]
wait so you deleted conda
```
(base) d3k768@WE47574 ~ % rm -r /Users/d3k768/mambaforge3
```

then this doesnt work
```
(base) d3k768@WE47574 ~ % conda init --all
```

wow

**Reply by GiudGiud** [reply]
(we wanted you to re-install conda after deleting it). see what Jason said

> Regardless, once you remove the directory above, please follow our [Conda Installation Instructions](https://mooseframework.inl.gov/getting_started/installation/conda.html) to obtain the latest version available.

**Reply by milljm** [reply]
> wow

Yeah, even with Conda removed, the initialization function is ever present.

@styyokuda:
Once you delete that directory, try closing the terminal, and reopening it so that function cannot load. This may be a little off topic, but this is what is doing it (mostly likely somewhere in your `~/.zshrc` file):

```bash
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/d3k768/mambaforge3/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/d3k768/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/d3k768/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/d3k768/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```
This effectively creates a function called `conda`. But only if the directory exists. So I suspect now that `mambaforge3` is missing, the above logic won't fire.

Please now proceed to install Conda using our instructions: [Conda Installation Instructions](https://mooseframework.inl.gov/getting_started/installation/conda.html)

The problem I am worried about, is when you do go to run `conda init --all`, it may not do anything. Because Conda may believe there is nothing to do since it may false positively detect `#>>> conda initialize` functions already present...

That is what I meant by "we will deal with that when/if it happens".

**Reply by styyokuda** [reply]
Thank you very much all, 
So, I repeated the instructions from the "Conda Installation Instructions" link again. Now I got the following:

d3k768@WE47574 ~ % curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-arm64.sh
bash Miniforge3-MacOSX-arm64.sh -b -p ~/miniforge
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 57.1M  100 57.1M    0     0  40.1M      0  0:00:01  0:00:01 --:--:--  111M
ERROR: File or directory already exists: '/Users/d3k768/miniforge'
If you want to update an existing installation, use the -u option.

**Reply by GiudGiud** [reply]
delete the folder before redownloading. or use what you downloaded already

**Reply by styyokuda** [reply]
All,
I am still having the exact same problem with "conda init --all" after "export PATH=$HOME/miniforge/bin:$PATH" as before. Please help.

**Reply by milljm** [reply]
Can you paste the error? Is it really this? Or some other path?
```bash
(base) d3k768@WE47574 ~ % conda init --all
exit
__conda_exe:1: no such file or directory: /Users/d3k768/mambaforge3/bin/conda

Saving session...
...copying shared history...cp: /Users/d3k768/.zsh_history: Permission denied
```

**Reply by styyokuda** [reply]
Thank you.  Here it goes:

d3k768@WE47574 ~ % conda init --all
exit
no change     /Users/d3k768/miniforge/condabin/conda
no change     /Users/d3k768/miniforge/bin/conda
no change     /Users/d3k768/miniforge/bin/conda-env
no change     /Users/d3k768/miniforge/bin/activate
no change     /Users/d3k768/miniforge/bin/deactivate
no change     /Users/d3k768/miniforge/etc/profile.d/conda.sh
no change     /Users/d3k768/miniforge/etc/fish/conf.d/conda.fish
no change     /Users/d3k768/miniforge/shell/condabin/Conda.psm1
no change     /Users/d3k768/miniforge/shell/condabin/conda-hook.ps1
no change     /Users/d3k768/miniforge/lib/python3.10/site-packages/xontrib/conda.xsh
no change     /Users/d3k768/miniforge/etc/profile.d/conda.csh
modified      /Users/d3k768/.bash_profile
modified      /Users/d3k768/.zshrc
modified      /Users/d3k768/.config/fish/config.fish
modified      /Users/d3k768/.xonshrc
modified      /Users/d3k768/.tcshrc

==> For changes to take effect, close and re-open your current shell. <==


Saving session...
...copying shared history...cp: /Users/d3k768/.zsh_history: Permission denied

...saving history...shell_session_save_history:14: permission denied: /Users/d3k768/.zsh_history
truncating history files...shell_session_save_history:21: can't read history file /Users/d3k768/.zsh_history

[Process completed]

**Reply by milljm** [reply]
okay, this all well and good.

Can you cat the following two files, I would like to know what changes it did make:
```bash
cat /Users/d3k768/.zshrc
```
and then this one:
```bash
cat /Users/d3k768/.bash_profile
```
Also, what does this report:
```bash
echo $0
```

**Reply by styyokuda** [reply]
Thank you! Here they go:

(base) d3k768@WE47574 ~ % cat /Users/d3k768/.zshrc
# set ls aliases
alias ll='ls -alF'
alias la='ls -A'
alias l='ls -CF'

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/d3k768/miniforge/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/d3k768/miniforge/etc/profile.d/conda.sh" ]; then
        . "/Users/d3k768/miniforge/etc/profile.d/conda.sh"
    else
        export PATH="/Users/d3k768/miniforge/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
mamba activate moose

(base) d3k768@WE47574 ~ % 
(base) d3k768@WE47574 ~ % cat /Users/d3k768/.bash_profile

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/d3k768/miniforge/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/d3k768/miniforge/etc/profile.d/conda.sh" ]; then
        . "/Users/d3k768/miniforge/etc/profile.d/conda.sh"
    else
        export PATH="/Users/d3k768/miniforge/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<

(base) d3k768@WE47574 ~ % 
(base) d3k768@WE47574 ~ % echo $0
-zsh
(base) d3k768@WE47574 ~ %

**Reply by milljm** [reply]
According to your original post you agreed that you would:

> I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

Can you edit the above message and enclose the output with the appropriate markdown, so its easier to read?

**Reply by styyokuda** [reply]
I apologize that I did not understand what I was asked. So, I am trying now.
```
(base) d3k768@WE47574 ~ % cat /Users/d3k768/.zshrc
# set ls aliases
alias ll='ls -alF'
alias la='ls -A'
alias l='ls -CF'

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/d3k768/miniforge/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/d3k768/miniforge/etc/profile.d/conda.sh" ]; then
        . "/Users/d3k768/miniforge/etc/profile.d/conda.sh"
    else
        export PATH="/Users/d3k768/miniforge/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
mamba activate moose
```
```
(base) d3k768@WE47574 ~ % cat /Users/d3k768/.bash_profile

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/d3k768/miniforge/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/d3k768/miniforge/etc/profile.d/conda.sh" ]; then
        . "/Users/d3k768/miniforge/etc/profile.d/conda.sh"
    else
        export PATH="/Users/d3k768/miniforge/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```
```
(base) d3k768@WE47574 ~ % echo $0

-zsh
```

**Reply by styyokuda** [reply]
Do you suggest the full re-install to make it happen?
I don't know if the full re-install helps but please let me know.
Please help.

**Reply by styyokuda** [reply]
All,
I found the following two files in the main directory:
```
Mambaforge-MacOSX-arm64.sh
Miniforge3-MacOSX-arm64.sh
```
Are these causing any problem? 
I need to get the project go on.
Please help. S. Thomas

**Reply by styyokuda** [reply]
All,
My ".bash_profile" and ".zshrc" files seemingly need to be updated and I don't know how.
Could someone send me these files so that I can update them?
I need to make the project progress.
Please help, S. Thomas

**Reply by GiudGiud** [reply]
Hello

The .sh scripts for both versions of conda do not cause any trouble.

If you have run both and have two conda on the machine that is a problem

For the bash_profile, first, are you using bash or zsh?
Echo $0 will show what you are using

**Reply by styyokuda** [reply]
I am using the "zsh" shell.
Thank you very much for your time and help.

**Reply by GiudGiud** [reply]
Ok then do not worry about the bash_profile

You are write that conda gets confused 

Conda init --all 
Should have initialized for the zsh shell too
If not look into the manual for conda init

**Reply by styyokuda** [reply]
I still have the same problem as before (please see below).
It reads "permission denied:  /Users/d3k768/.zsh_history"
I cannot open the ".zsh_history" file. 
Is this my computer problem?
Please help, S. Thomas

```
(base) d3k768@WE47574 ~ % export PATH=$HOME/miniforge/bin:$PATH
(base) d3k768@WE47574 ~ % conda init --all
exit
no change     /Users/d3k768/miniforge/condabin/conda
no change     /Users/d3k768/miniforge/bin/conda
no change     /Users/d3k768/miniforge/bin/conda-env
no change     /Users/d3k768/miniforge/bin/activate
no change     /Users/d3k768/miniforge/bin/deactivate
no change     /Users/d3k768/miniforge/etc/profile.d/conda.sh
no change     /Users/d3k768/miniforge/etc/fish/conf.d/conda.fish
no change     /Users/d3k768/miniforge/shell/condabin/Conda.psm1
no change     /Users/d3k768/miniforge/shell/condabin/conda-hook.ps1
no change     /Users/d3k768/miniforge/lib/python3.10/site-packages/xontrib/conda.xsh
no change     /Users/d3k768/miniforge/etc/profile.d/conda.csh
no change     /Users/d3k768/.bash_profile
no change     /Users/d3k768/.zshrc
no change     /Users/d3k768/.config/fish/config.fish
no change     /Users/d3k768/.xonshrc
no change     /Users/d3k768/.tcshrc
No action taken.

Saving session...
...copying shared history...cp: /Users/d3k768/.zsh_history: Permission denied

...saving history...shell_session_save_history:14: permission denied: /Users/d3k768/.zsh_history
truncating history files...shell_session_save_history:21: can't read history file /Users/d3k768/.zsh_history

[Process completed]
```

**Reply by GiudGiud** [reply]
That zsh_history file is the history of all commands run and should not matter. 
You could try to delete it if its permissions are not right

It says no change in your .zshrc 

So is the conda snippet in that file already?

**Reply by styyokuda** [reply]
Thank you for your suggestion.  I moved the .zsh_history file to .zsh_history_back. Then, I was able to do the following procedure given in the Conda Installation Instructions.  Now can I repeat the update procedure?

```
conda update --all --yes
```
```
conda config --add channels https://conda.software.inl.gov/public
```

**Reply by styyokuda** [reply]
Sorry I forgot to show my .zshrc file in the previous (above) post.  Here it is:
```
(base) d3k768@WE47574 ~ % cat .zshrc
# set ls aliases
alias ll='ls -alF'
alias la='ls -A'
alias l='ls -CF'

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/d3k768/miniforge/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/d3k768/miniforge/etc/profile.d/conda.sh" ]; then
        . "/Users/d3k768/miniforge/etc/profile.d/conda.sh"
    else
        export PATH="/Users/d3k768/miniforge/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
mamba activate moose
```

**Reply by GiudGiud** [reply]
this looks fine. Please try to update if that was the goal

**Reply by milljm** [reply]
This all looks good to me! You should be able to create the moose conda environment:
```bash
conda create -n moose moose-dev
```
That last line your `~/.zshrc` should be changed from mamba to conda:
```diff
unset __conda_setup
# <<< conda initialize <<<
- mamba activate moose
+ conda activate moose
```

**Reply by styyokuda** [reply, marked as answer]
Thank you very much, all!
It is working now.  The key point was to move the .zshrc_hisotry file to .zshrc_history_back because I did not have  permissions for anything else, vim, rm, etc.
Again, thank you very much for your time and patience with me.  I will try to update as soon as possible as recommended because waiting too long seems to cause a problem(s).
S. Thomas
